### DISCUSSION TITLE
useAtom doesn't work for atomWithStorage - React Native/AsyncStorage

### DISCUSSION THREAD

**Discussion by francoascarate-igz** [pmndrs/jotai#1662] [Q&A]
## Summary

I have this function to create `asyncStorage atoms`
```
export function createAsyncStorageStore<T>(key: string, initialState: T) {
  const storage = createJSONStorage<any>(() => AsyncStorage);
  const storedAtom = atomWithStorage<T>(key, initialState, storage);

  return storedAtom;
}
```

When I try to use the atoms created by this function like this:
```
const USER_ATOM_KEY = 'user';
export const userAsyncStorageAtom = createAsyncStorageStore<boolean>(
  USER_ATOM_KEY,
  false,
);

export function useStore<T>(store: PrimitiveAtom<T>) {
  return useAtom(store);
}

const [userAsyncStorage, setUserAsyncStorage] = useStore(userAsyncStorageAtom);
```
I get this TypeScript error:
```
Argument of type 'WritableAtom<Promise<boolean>, SetStateActionWithReset<boolean>, Promise<void>>' is not assignable to parameter of type 'PrimitiveAtom<boolean | unique symbol>'.
  Type 'Promise<boolean>' is not assignable to type 'boolean | unique symbol'.
```

## Link to reproduction
[Typescript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgQxhEAaOBXAzgUwEE1M4AFKUYGYANyJLgF84AzKdOAcgCsIZkwLgG4AUKEixEcAMZR8qfACkAygHkAcirRRkAc3xZU6AOrUAFtuj78zNhxDc+A4AHpsNADa4u48NHhCXABPADsZK10De04uAAF5ZBkYAFpQ1Dp8FOQQ8JTcHRtXHLCZfMKDX1F8AA9JeFZscJoIUNlEmCJciIr8SPwAHgAVAD4ACgBrfGCALjgCylC9LGBQ6mBkT21FOaGASkRROFlWgvneuABedoVO1U1ImwHkUODxsYPLkbgg0seDPZiY4yU7wArQfAAE2InGuxhAZhgll6w3GU2CKzWNE2206WHBUXwgNERzg8hg2CgbQJUJhIDETBJtXqbCayWArRuin6qLGq3WOIEnV2BwQpPJlLa8N5-OxWyFRIZJOqdQCrOaHLaeD6OkGozGNLmFCoNHodNRovF+ApVJwBDpBt1xMZohBoTOAFUVABRABKAH1CEM1ABZf0AaW9AE0rtxtVAREzVVI3Wd479wv8GLCuZ0Mz1rAYeQAjCAQTwKULjUlev2B4NhyNRjCk1ibAgt4kqlmp+DpxjXOS3HUQgal8uV8Zt7yKkm9uAAbXT3Sz+OtHoIUHzWYAurHtf0xsu-r06cT50vN2uYBv8FA99cD7qj5uz2IgA)

- How can I make it work without using any to get the types?
- This works with the normal atoms (example in Typescript playground)
## Check List

Please do not ask questions in issues.

- [x] I understand this is not a question.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
It's because what you create is not a primitive atom:

Here's a fix: https://tsplay.dev/WYR1rw
